### PR TITLE
[WIP] Generalized Temporal Graph Definition

### PR DESCRIPTION
**Pull request by rwoll**
@ericthewry @MelanieWu 

1. I have added a `builds` directory in the root of the project so you can easily read the PDFs without
    having to clone and compile on your machine. (I know this is not necessarily best practice since the
    PDF could get out of sync with the `.tex` file, but I figured it was convenient.)

2. As you will see in the diff below/above I modified the definitions to use a more generalize, all-
    encompassing temporal graph definition which corresponds to our discussion on Friday with Prof.
    Wu. I also added some general terms and their meaning since I figured it would be convenient to have
    everything in one place and reworded some of Eric's definitions in terms of those.

It's simply a quick revision, so please feel free to comment with corrections/discussion points/reversions.

Best,
@rwoll

### CODE REVIEW

**Comment by ericthewry** [comment]
Excellent. I mentioned this before, but this seems like a very object-oriented approach to definitions, something that I've not generally seen in Graph Theory. It might be appropriate for our purposes though. Generally, graph theorists like to think about different kinds of graphs as discrete structures with various isomorphisms to convert between them, that  are semantically approximate, but result in different properties. More complicated graphs are thought of as special cases of the simple graph instead of vice-versa.  This is how `graph-tools` defines their graphs as well. Doesn't largely matter, just a style choice I guess.

These definitions are (mostly) consistent with the theory.  I've made comments throughout.

File: paths.tex [hunk 1]
```diff
@@ -13,48 +15,118 @@
...
-and address their usefulness by examining algorithms and attempting to prove
-different facts.
+\begin{defn}
+  A \textbf{temporal graph} or \textbf{temporal network}\footnote{We will treat
+  the terms \textit{graph} and \textit{network} synonymously throughout this
+  paper.} is defined as a tuple $G \in V \times E$ where $V$ is the set of
+  vertices, and $E \in V^2 \times T^2$ is the set of edges such that
+  $T = [0, \infty)$:
+
+  $$G = (V, E)$$
```
Comment (ericthewry) [inline paths.tex]: Entry into mathmode via `$$` is depecated. It is standard to use
```
  \[ G = (V,E) \]
```

Technically, single dollar sign is also deprecated, with preference for `\( T = [0, \infty] \)`, but I've never seen that used.

File: paths.tex [hunk 2]
```diff
@@ -13,48 +15,118 @@
...
+  are \textit{active}.
+\end{defn}
+
+With this generalized definition of a temporal network, we will begin to explain
+some common, specialized descriptors of these networks:
+
+\begin{defn}
+  We will describe a temporal graph as \textbf{undirected} $\forall e \in E$
+  if there exists $e' \in E$ such that $e = (v_i, v_j, t_0, t_f,w)$ and $e' =
+  (v_j, v_i, t_0, t_f, w)$.\footnote{As we consider the real-life applications
```
Comment (ericthewry) [inline paths.tex]: This definition of undirected is not really accurate, and will call for redefinition of lemmas and theorems (Handshake Lemma included).  This definition also has some unintended structural properties, where there is a two-cycle from $u$ to $u$ for all $u \in V$ (in a graph $(V,E)$). 

It also has some unintended consequences for trails -- allowing double traversal for an edge. A trail is a walk with no repeated edge. In an undirected graph with this definition, for every edge (u,v,...) you can have a trail that picks up lone vertices. E.g. by your definition, there is a closed trail that connects all of these, A, B, D, C, A, where in the standard understanding of an undirected graph there is no such trail, because each edge can be traversed only once.
```
(A)
    \
    (B)  -  (C)
      \_(D)_/
```

This standard way of converting between directed and undirected graphs is okay for paths, cycles, walks, connectivity, and many other things, but falls apart in other areas, so is not ideal as a base definition.

Standardly, a directed edge is defined as an ordered pair of vertices, and an undirected edge is defined as an unordered pair of vertices. This becomes slightly inelegant when the edge becomes a 5-tuple, but I think is necessary to avail ourselves of much of the undirected theory.

File: paths.tex [hunk 3]
```diff
@@ -13,48 +15,118 @@
...
+  of these temporal networks, it may be convenient to define an edge as
+  a 6-tuple with the last component $l$ being a unique label or edge
+  differentiator. This would allow for the following scenario: if author
```
Comment (ericthewry) [inline paths.tex]: Not sure where this part fits in (about the differentiator). It feels like a valuable inclusion as an implementation detail, but is usually not included in mathematical definitions.  This is handled by saying "Let e_1 and e_2 be two distinct edges from $u$ to $v$ in window $[t_1 , t_2]$ with weight $w$".

File: paths.tex [hunk 4]
```diff
@@ -13,48 +15,118 @@
...
+  \texttt{A} and \texttt{B} collaborate in window $[t_0, t_f]$ on multiple
```
Comment (ericthewry) [inline paths.tex]: Use \emph or mathmode.

File: paths.tex [hunk 5]
```diff
@@ -13,48 +15,118 @@
...
-  When the vertex pair within the edges is ordered, call the graph
-  \textbf{directed}; otherwise call it \textbf{undirected}.
+  A \textbf{instantaneous edge} or \textbf{contact edge} is any edge $e \in E$
+  where $t_0 = t_f$.
 \end{defn}
 
-Specifically we will analyze the effect of these definitions on the coauthorship
-and citation networks.
+\begin{defn}
+  \textbf{Interval networks} or \textbf{windowed networks} shall be defined as
```
Comment (ericthewry) [inline paths.tex]: We need to be careful with terminology here.  An interval graph is a static structure that is often used in a very different way than we are describing (its something I've been intending to bring up but haven't since it hasn't been pressing, might be useful for our definitions of contemporaneity). Since you above said that network and graph are interchangable, we'll get into trouble when we want to introduce these interval graphs.

File: paths.tex [hunk 6]
```diff
@@ -13,48 +15,118 @@
...
+  every edge in the network is an \textit{instantaneous edge} or \textit{contact
+  edge}. It is simply a special case of an interval network.
+\end{defn}
 
 \begin{defn}
-  A \textbf{coauthorship network} is an undirected edge-varying temporal network
-  in which each node represents an author, and the existence of an edge represents
-  a collaboration over a period of time.
+  An \textbf{aggregate network} shall be defined as a temporal network where
+  every edge is \textbf{infinitely persistent}.
```
Comment (ericthewry) [inline paths.tex]: This makes an aggregate network a sub-case of the temporal network, and I see it as 
a completely different structure.  It doesn't vary over time, and so it seems like an unnecessary step of complexity.

It is a correct definition, but not one that I have seen in any of the other literature.

File: paths.tex [hunk 7]
```diff
@@ -150,29 +239,30 @@ \subsection{Infinitely Persistent Edges}
...
-  degree 2.  Most importantly, $t_i \geq s_{i+1}$ for all $i \in [n-2]$.
+  edges $(v_1,v_2,s_1,t_1),(v_2,v_3,s_2,t_2), \cdots (v_{n-1},v_{n}, s_{n-1},
+  t_{n-1})$ such that $v_1$ and $v_n$ have degree $n$, and the remaining
+  vertices have degree 2.  Most importantly, $t_i \geq s_{i+1}$ for all $i \in
+  [n-2]$.
 \end{defn}
 
 The definitions for fastest and shortest path will be the same as in definition
-2.1.2.
+[?].
```
Comment (ericthewry) [inline paths.tex]: maybe use a \ref here?

File: .gitignore [hunk 8]
```diff
@@ -0,0 +1,175 @@
+# https://github.com/github/gitignore
```
Comment (ericthewry) [inline .gitignore]: :D

File: paths.tex [hunk 9]
```diff
@@ -13,48 +15,118 @@
...
+  paper.} is defined as a tuple $G \in V \times E$ where $V$ is the set of
+  vertices, and $E \in V^2 \times T^2$ is the set of edges such that
+  $T = [0, \infty)$:
+
+  $$G = (V, E)$$
+
+  Each edge $e \in E$ is a 5-tuple $(v_i, v_j, t_0, t_f, w)$ such that
+  $v_i, v_j \in V$, $t_0, t_f \in T$, and $t_0 \leq t_f$. An edge is said to
+  connect $v_i$ to $v_j$ (with the possibility of $v_i = v_j$) between the time
+  $[t_0, t_f]$ with weight $w$. That is, all edges are to be treated as directed
```
Comment (ericthewry) [inline paths.tex]: domain of w?

**Comment by ericthewry** [comment]
also, this diff is really weird... poor git.

File: paths.tex (hunk 7 above)
Comment (rwoll) [inline paths.tex]: :+1: Once everything's stabalized a bit, I will go back through and add it.

File: paths.tex (hunk 1 above)
Comment (rwoll) [inline paths.tex]: Noted.

File: paths.tex (hunk 9 above)
Comment (rwoll) [inline paths.tex]: Should we abstractly define the domain as `W` or be more precise and deem it must be rational? My inclination is to leave it open and let it be specified on a topic-per-topic basis with most cases using rationals.

What are your thoughts?

File: paths.tex (hunk 2 above)
Comment (rwoll) [inline paths.tex]: Yes -- we probably want to discuss this further and consider the implications.

For theory I think it makes most sense as an unordered pair, but in application it is convenient to define an edge in both directions and adapt the logic and other theorems to avoid the caveats that you pointed out. I'm not sure which is ultimately better. I like the idea of having the generic definition and this allows for that, but also understand it create complications elsewhere.

File: paths.tex (hunk 3 above)
Comment (rwoll) [inline paths.tex]: :+1: I will remove that footnote in the next iteration. I threw myself off by not specify that `E = {e_0, e_1, ..., e_n}`.

File: paths.tex (hunk 5 above)
Comment (rwoll) [inline paths.tex]: I simply meant to create a distinction between a temporal graph whose edges have intervals of times versus the special case of this where the intervals are constant. Would you mind elaborating on what you mean?

File: paths.tex (hunk 6 above)
Comment (rwoll) [inline paths.tex]: I see it as varying over time making it suited as a sub-case. An aggregate network (at least the way defined in this document) allows for varying start times on each of the edges; therefore, if you look at an aggregate network in year `t_n` and then again in `t_{n+1}` it may have changed and therefore varies in time. Edges cannot become "de-activated", but they come into "activation" at different times. Perhaps we may want to choose a different word to refer to this behavior.

**Comment by rwoll** [comment]
@ericthewry I think some of the altered line breaks made the diff less than ideal; any recommendations around this?

**Comment by ericthewry** [comment]
I have no idea:/

File: paths.tex (hunk 3 above)
Comment (ericthewry) [inline paths.tex]: Thats actually also unnecessary. Also, super pedantically, $|E| = m$ and $|V| = n$

File: paths.tex (hunk 5 above)
Comment (ericthewry) [inline paths.tex]: From Adam West:
"An interval representation of a graph is a family of intervals assigned to the vertices so that vertices are adjacent if and only if the corresponding intervals intersect. A graph having such a representation is an interval graph."

This is a tool that we could use to analyze temporal communities, but it is something that is considerably different from what you are referring to here.

File: paths.tex (hunk 6 above)
Comment (ericthewry) [inline paths.tex]: That's a distinction I made in the original document.  An aggregate network removes all notion of time.  What you have defined is a temporal graph with infinite edge persistence. And these are different things.

File: paths.tex (hunk 9 above)
Comment (ericthewry) [inline paths.tex]: hmm. i'm not sure what's best here... @MelanieWu ?

File: paths.tex (hunk 2 above)
Comment (ericthewry) [inline paths.tex]: I agree that it is a good implementational heuristic. Since this is currently a set of theoretical definitions, I say we go with the standard, theoretical definition. When we move to implementation, we can show that this theoretical definition will have to argue that this representation doesn't affect notions of centrality/community in bizarre ways. This is a non trivial fact, that will require only one proof, where changing the definition this early will require a new proof everytime we want to use an existing theorem.

File: paths.tex (hunk 6 above)
Comment (rwoll) [inline paths.tex]: Then perhaps my definition is better described as an aggregat*ing* network. And an aggregate network describes your original definition -- essentially an aggregating graph at the maximum at the point in time in which no new edges are formed/forming.

Comment (rwoll) [inline paths.tex]: Also, the definition in this commit was meant to define something different than the aggregated *static* network described by [Nicosia et. al](http://arxiv.org/pdf/1306.0493v1.pdf).

**Comment by rwoll** [comment]
@ericthewry Commit 73699b8 contains patches from both our on- and off-line conversations. 

Things still under immediate consideration (thus left unchanged):
- To define or not define: edge weight domain
- Undirected representation: using an unordered tuple definition versus the directed special case definition

Todos:
- Label figures
- adding `\ref{...}`s

**Comment by ericthewry** [comment]
@rwoll Commit 62ac78c contains a definition and transformation description that keeps the implementational details you liked, and allows us to be consistent with undirected theory. Commit c8315fc contains some LaTeX fixes.

I also decided to bind $W$, limiting it to some numeric set $W \subseteq \mathbb{R}$. The definition i've seen for weightedness is to define a weighted graph as $G = (V, E, f)$ where $f$ is a function from $E$ to numeric values.  I like our tuple definition better.

**Comment by ericthewry** [comment]
@MelanieWu This has reached a point where your feedback on the definitions would be greatly appreciated:)

**Comment by ericthewry** [comment]
@rwoll Have a look when you get a chance. I'm going to merge (into the original PR) by noon tomorrow if I don't hear anything, just so I can keep making progress and avoid unnecessary PR bloat.  Once its merged, feel free to make additional PRs if you find something that needs changing.

File: paths.tex [hunk 10]
```diff
@@ -255,39 +252,26 @@ \subsection{Windowed edges}
...
-\end{defn}
-
-The definitions for fastest and shortest path will be the same as in definition
-\ref{defn:short_fast_path}.
+The definitions for temporal paths will be the same as for infinitely persistent
+edges, \ref{defn:cons_temp_path}. As well as those for fastest and shortest path
+will be the same as in definition \ref{defn:short_fast_path}.
 
-\section{Pairwise Contemporaneity}
+\section{Complete Contemporaneity}
```
Comment (rwoll) [inline paths.tex]: Perhaps we should call this *Concurrent Contemporaneity* -- for some reason my brain isn't reading *complete* in its intended way.

File: paths.tex [hunk 11]
```diff
@@ -44,44 +49,36 @@ \section{Preliminary Definitions}
...
+  $\T$ is a function $\T : E \to \Z^2$ with the co-domain being the start and
+  end times of the vertices, in number of time units (you can always pick a
+  smaller time unit so that the bounds of this interval are integral). Note that
+  forall $e \in E$ with $\T(e) = (t_s, t_f)$, $t_s \leq t_f$. In a
+  \textbf{static} graph, where $\T(e) = (\infty, \infty)$ for all $e \in E$,
+  leave $\T$ out of the definition of the graph. $\W$ is a function
+
+  $\W : E \to \Z$ representing the integer weight of each edge. In an
+  \textbf{unweighted} graph if every edge has the same weight, then we can
+  leave $\W$ out of the definition.
```
Comment (rwoll) [inline paths.tex]: Good to formally state this.

File: paths.tex [hunk 12]
```diff
@@ -44,44 +49,36 @@ \section{Preliminary Definitions}
...
-  existence of the other.
-\end{defn}
-
 Note that there is a direct transformation between undirected temporal graphs and
 directed temporal networks. To go from undirected to directed, for every edge
 $(e_l,e_r) \in E^2$, add edges $e_l \in E$ and $e_r \in E$ to the directed
 network. To go from directed to undirected, if for every $u,v$ pair in the
 directed network, the number of edges of the form $(u,v,t_0,t_f,w)$ is the same
-as the number of edges of the form $(u,v,t_0,t_f,w)$, then pair up these edges;
+as the number of edges of the form $(v,u,t_0,t_f,w)$, then pair up these edges;
```
Comment (rwoll) [inline paths.tex]: Should we change this to use the functional definitions?

File: paths.tex [hunk 13]
```diff
@@ -153,10 +149,11 @@ \section{Preliminary Definitions}
...
- A \textbf{shortest path} between $v_1, u$ (di)graph $G = (V,E)$ is
- $P = v_1, v_2, \cdots, v_n, u$ such that $v_i \in V$ for all $i \in [n]$,
- and $v_jv_{j+1}, v_nu \in E$, for all $j \in [n-1]$, and there is no path
- $P' = v_1, v_2, \cdots, v_m, u$ such that $m < n$.
+ A static \textbf{shortest path} between $u,v$ in a static (di)graph
+ $G = (V,E,\W)$ is $P = e_1 = (u,u'), e_2, \cdots, e_n = (v',v)$,
+ $P \in E$ such that for all $e_i = (v_1,v_2), e_{i+1} = (v_3,v_4) \in E$,
+ $v_2 = v_3$, and for every $u,v$-path $P' = e'_1, e'_2, \cdots, e'_m$
+  \[\sum_{e'_i \in P'} \W(e'_i) < \sum_{e_i \in P} \W(e_i) \]
  \end{defn}
```
Comment (rwoll) [inline paths.tex]: :+1: Disambiguates path

File: paths.tex [hunk 14]
```diff
@@ -153,10 +149,11 @@ \section{Preliminary Definitions}
...
```
Comment (rwoll) [inline paths.tex]: :+1: Disambiguates each path using edges

File: paths.tex [hunk 15]
```diff
@@ -169,11 +166,10 @@ \section{Consecutive Contemporaneity}
...
 \begin{defn}
-  A \textbf{consecutive temporal path} between $v_1$ and $v_n$ is a path
-  $P = v_1, v_2, \cdots, v_n$ such that $v_i \in V$ for all $i \in [n]$,
-  and $v_jv_{j+1} \in E$, for all $j \in [n-1]$, and for every pair of
-  consectutive edges $(v_{i-1},v_{i})_{t_1}$ and $(v_{i}, v_{i+1})_{t_2}$,
-  $t_1 \leq t_2$.
+  \label{defn:cons_temp_path}
+  A \textbf{consecutive temporal path} between $u$ and $v$ is a $v_1,v_n$-path
+  $P = e_1, e_2, \cdots, e_n$ such that for consecutive edges $e_i,e_{i+1} \in P$,
+  $T(e_i) \cap T(e_j) \neq \emptyset$.
```
Comment (rwoll) [inline paths.tex]: Should we add in the generalize minimum shape graphs for clarity? Or at the very least some of the graphs we were drawing on the board -- they are an easy way to see the subtle differences in the types of paths we are defining.

Comment (ericthewry) [inline paths.tex]: I'm not really sure how to give a formal notion of this.  Only an intuitive one. If you have thoughts for what that should look like, maybe make another PR?

**Comment by ericthewry** [comment]
@rwoll I made those edits in the last commit.  Taking care of the changes to the definitions of path in PR #5 . Should be good to go.

File: paths.tex [hunk 16]
```diff
@@ -77,9 +77,9 @@ \section{Preliminary Definitions}
 directed temporal networks. To go from undirected to directed, for every edge
 $(e_l,e_r) \in E^2$, add edges $e_l \in E$ and $e_r \in E$ to the directed
 network. To go from directed to undirected, if for every $u,v$ pair in the
-directed network, the number of edges of the form $(u,v,t_0,t_f,w)$ is the same
-as the number of edges of the form $(v,u,t_0,t_f,w)$, then pair up these edges;
-otherwise there is no transformation.
+directed network, the number of edges of the form $e = (u,v)$ is the same
+as the number of edges of the form $e' = (v,u)$, with $\T(e) = \T(e')$ and
+ $\W(e) = \W(e')$ then pair up these edges; otherwise there is no transformation.
```
Comment (rwoll) [inline paths.tex]: Noted. Thanks for adjusting these.

File: paths.tex [hunk 17]
```diff
@@ -77,9 +77,9 @@ \section{Preliminary Definitions}
 directed temporal networks. To go from undirected to directed, for every edge
 $(e_l,e_r) \in E^2$, add edges $e_l \in E$ and $e_r \in E$ to the directed
 network. To go from directed to undirected, if for every $u,v$ pair in the
-directed network, the number of edges of the form $(u,v,t_0,t_f,w)$ is the same
-as the number of edges of the form $(v,u,t_0,t_f,w)$, then pair up these edges;
-otherwise there is no transformation.
+directed network, the number of edges of the form $e = (u,v)$ is the same
```
Comment (rwoll) [inline paths.tex]: Looking over this again, I'm not sure the transformation stated like this makes sense since the the graph is a multi-graph; given nodes *a* and *b*, there could exist *a*->(1990)*b* and *b*->(1990)*a* such that these two edges were not generated from the same entity (i.e. this is a contextual definition). What are your thoughts? @ericthewry

Comment (ericthewry) [inline paths.tex]: I'm not partial to this section. Its not important to anything we want to do. We'll more usually just use the underlying graph when we want an undirected from a directed. It was intended to give intuition about the old definition we had. Now we dont need it.

Comment (ericthewry) [inline paths.tex]: I don't really understand what you mean about the contextual definition. But I agree for different reasons that this is not a useful section

Comment (rwoll) [inline paths.tex]: *Contextual* as in it applies to certain directed graphs depending on how they were generated. For instance, our current DBLP databases graph has all edges duplicated so this definition applies, but for other graph sets/data-representations it becomes incorrect. Either way, since we both think it is not very helpful, we might want to consider deleting it from the document.

**Comment by ericthewry** [comment]
@rwoll  Removed that last piece. good to merge?

**Comment by rwoll** [comment]
@ericthewry Good to merge. If you are happy with it as well, go ahead and merge so we can focus on #5 . (If you want me to squash some of the commits and submit another PR, I could do that as well.)

**Comment by ericthewry** [comment]
@rwoll Go ahead on the PR if you have some thoughts for #3
